core/src/main/java/org/elasticsearch/script/ScriptMetaData.java

```diff
@@ -18,168 +18,414 @@
+            out.write(scripts.size());
+
+            for (Map.Entry<String, StoredScriptSource> entry : scripts.entrySet()) {
+                out.writeString(entry.getKey());
+                entry.getValue().writeTo(out);
+            }
+        // Prior to Version 5.2, stored scripts can only be read using the deprecated
+        // namespace.  Scripts using the deprecated namespace are first isolated in a
+        // temporary Map, then written out.  Since all scripts will be stored using the
+        // deprecated namespace, no scripts will be lost.
+        } else {
+            Map<String, StoredScriptSource> filtered = new HashMap<>();
+
+            for (Map.Entry<String, StoredScriptSource> entry : scripts.entrySet()) {
+                if (entry.getKey().contains("#")) {
+                    filtered.put(entry.getKey(), entry.getValue());
+                }
+            }
+
+            out.writeVInt(filtered.size());
+
+            for (Map.Entry<String, StoredScriptSource> entry : filtered.entrySet()) {
+                out.writeString(entry.getKey());
+                entry.getValue().writeTo(out);
+            }
+        }
     }
 
+    /**
+     * This will parse XContent into {@link ScriptMetaData}.
+     *
+     * The following format will be parsed for the new namespace:
+     *
+     * {@code
+     * {
+     *     "<id>" : "<{@link StoredScriptSource#fromXContent(XContentParser)}>",
+     *     "<id>" : "<{@link StoredScriptSource#fromXContent(XContentParser)}>",
+     *     ...
+     * }
+     * }
+     *
+     * The following format will be parsed for the deprecated namespace:
+     *
+     * {@code
+     * {
+     *     "<id>" : "<code>",
+     *     "<id>" : "<code>",
+     *     ...
+     * }
+     * }
+     *
+     * Note when using the deprecated namespace, the language will be pulled from
+     * the id and options will be set to an empty {@link Map}.
+     */
     @Override
     public ScriptMetaData fromXContent(XContentParser parser) throws IOException {
-        Map<String, ScriptAsBytes> scripts = new HashMap<>();
-        String key = null;
+        Map<String, StoredScriptSource> scripts = new HashMap<>();
+        String id = null;
+        StoredScriptSource source;
+
         for (Token token = parser.nextToken(); token != Token.END_OBJECT; token = parser.nextToken()) {
             switch (token) {
                 case FIELD_NAME:
-                    key = parser.currentName();
+                    id = parser.currentName();
                     break;
                 case VALUE_STRING:
-                    scripts.put(key, new ScriptAsBytes(new BytesArray(parser.text())));
+                    if (id == null) {
+                        throw new ParsingException(parser.getTokenLocation(),
+                            "unexpected token [" + token + "], expected [<id>, <code>, {]");
+                    }
+
+                    int split = id.indexOf('#');
+
+                    if (split == -1) {
```

Comment:
I think we can move the split check to the place where we assign the `id` variable?